Debate/Cost.lean

```diff
@@ -109,6 +111,44 @@ theorem bob_debate_cost (o : Oracle) (alice : Alice) (vera : Vera) (t : ℕ):
   · refine exp_le_of_forall_le (fun y m ↦ ?_)
     induction y; repeat simp only [Comp.cost_pure, le_refl]
 
+/-- `Nat.ceil` adds at most one -/
+lemma Nat.ceil_le_add_one {x : ℝ} (x0 : 0 ≤ x) : ⌈x⌉₊ ≤ x + 1 := by
+  rw [natCast_ceil_eq_intCast_ceil x0]
+  exact (Int.ceil_lt_add_one x).le
+
+/-- Alice makes `O(k^2 t log t)` queries with default parameters -/
+theorem alice_fast (k : ℝ) (k0 : 0 < k) (t : ℕ) (bob : Bob) (vera : Vera) :
+    let p := defaults k t k0
+    (debate (alice p.c p.q) bob vera t).cost' o AliceId ≤
+      (t+1) * (5000 * k^2 * Real.log (200 * (t+1)) + 1) := by
+  refine le_trans (alice_debate_cost _ _ _ _) ?_
+  refine mul_le_mul_of_nonneg_left ?_ (by positivity)
+  simp only [defaults, samples, ← Real.log_inv]
+  generalize hn : (t+1 : ℝ) = n
+  field_simp
+  simp only [mul_pow, mul_div_assoc (Real.log _), mul_div_right_comm, mul_right_comm _ _ (2 : ℝ)]
+  norm_num
+  simp only [mul_comm (Real.log _)]
+  refine le_trans (Nat.ceil_le_add_one ?_) (le_refl _)
+  exact mul_nonneg (by positivity) (Real.log_nonneg (by linarith))
+
+/-- Bob makes `O(k^2 t log t)` queries with default parameters -/
+theorem bob_fast (k : ℝ) (k0 : 0 < k) (t : ℕ) (alice : Alice) (vera : Vera) :
+    let p := defaults k t k0
+    (debate alice (bob p.s p.b p.q) vera t).cost' o BobId ≤
+      (t+1) * (20000 / 9 * k^2 * Real.log (200 * (t+1)) + 1) := by
```

Comment:
Is it feasible to write a version of this using asymptotics notation, that more closely resembles the docstring?

Reply:
I want the explicit constant version: https://x.com/geoffreyirving/status/1805903946629701764

Reply:
I want the explicit constant version: https://x.com/geoffreyirving/status/1805903946629701764

Reply:
Ah, you probably mean *additionally* have the asymptotic versions. I’d like to triage that away for now if possible.